feat(lib): add real-valued implementation of `jax.scipy.special.fresnel`

jeertmans opened the pull request:
As discussed in #22683, I implemented a real-valued version of `jax.scipy.special.fresnel` by simply translating the original C++ code into JAX-compatible code. I haven't benchmarked the implementation (yet), nor tried to optimize it.

Reading the [contributing guide](https://jax.readthedocs.io/en/latest/contributing.html), I could not find much help on how to format my code, so I tried to stick to the same look as surrounding code.

Also, I wasn't sure if I should rather put the implementation inside the `lax` module or not, so please do not hesitate to comment on that.

Even if the implementation does not support complex, it achieves an accuracy that is extremely close to that of SciPy, and I think real-valued Fresnel integrals are quite useful, especially in radio propagation research tools.

Closes #22683

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -2653,3 +2653,216 @@ def log_softmax(x: ArrayLike,
+     The JAX version only supports real-valued inputs, and
+     is based on the SciPy C++ implementation, see
+     `here
+     <https://github.com/scipy/scipy/blob/1abc9faf43d7b82e0b26589bb705d8d009b18713/scipy/special/special/cephes/fresnl.h>`_.
```
If you based this on the scipy/cephes implementation, it will need to go into `jax/_src/third_party/scipy` along with the appropriate license

jakevdp replied:
Also, consider linking via a version tag rather than a commit hash

jeertmans replied:
Done!

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -2653,3 +2653,216 @@ def log_softmax(x: ArrayLike,
+
+  # This part is mostly a direct translation of SciPy's C++ code.
+
+  fresnl_sn = jnp.array([
```
Sometimes we've found that series expansions developed for 64-bit arithmetic fail in surprising ways when evaluated in 32-bit. Have you checked the precision of the outputs when `jax_enable_x64=False`?

jeertmans replied:
I have added a test and the output suggests that the series expansions work for both single and double precision.

jakevdp reviewed (commented):
Looks great! A couple comments below.

Also, please add the new function to the documentation listing here: https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipyspecial

jeertmans commented:
Thank you for your comments! I will address them next week :)

jeertmans commented:
I think I addressed all your comments @jakevdp, let me know it needs any additional change :-)

jakevdp commented on tests/lax_scipy_special_functions_test.py:
```diff
@@ -265,6 +268,18 @@ def testBetaParameterDeprecation(self):
       with self.assertRaises(TypeError), self.assertWarns(DeprecationWarning):
         lsp_special.beta(b=1, y=1)
 
+  def testFresnelSingleAndDoublePrecision(self):
```
What does this test cover that the `op_record` test above does not?

jeertmans replied:
I did that to check that single and double precision would both work, as you questioned, but I did not know wether the op_record would already test that.

jakevdp replied:
`op_record` already does everything that is in this test, I believe.

My comment wasn't really about test coverage, but rather a theoretical/algorithmic question. Do we expect this expansion to work in float32? Can you think of corner cases that are untested that might have issues?

jeertmans replied:
Indeed, it does. I will have the test removed by tomorrow then :)

For the cornecases, I don’t see any obvious one. But if you have an example from another function, maybe that could give us some ideas?

jeertmans replied:
I just removed the redundant test.

jakevdp replied:
Well, for example #17038 is currently blocked because the bessel function approximations don't converge in float32. This implementation looks quite similar to that, which is why I brought it up.

jeertmans replied:
I see no particular reasons that it might fail, and no documented use of a single precision version. If you prefer, I can also implement the [single-precision `fresnelf`](https://github.com/jeremybarnes/cephes/blob/master/single/fresnlf.c) from Cephes, if you think that is safer. Reading the implementation, it differs quite a bit from the double precision one.

I tried to understand how SciPy deals with `float32`, but it seems that they would just cast the data type to `float64`, but it isn't easy to be sure about what they actually do (at least to what I have found in their codebase).

jakevdp replied:
scipy generally deals with float32 by casting the inputs to float64, computing the function in float64, and then casting the outputs back to float32. That's not an option in JAX.

jeertmans replied:
Then, should I implement a `float32`-specific version to be sure that it is made for single precision?

jakevdp replied:
Maybe - it might be that the float64 expansion works just fine when cast to float32. But we'd need to explore that a bit to be sure. e.g. choose a range of inputs, compute the result in both float64 and float32, and confirm that the float32 output is within a reasonable range of the expected value given float32 precision.

jeertmans replied:
But wasn’t what my previous test was doing? I computed using float32 with JAX, and compared it to the float32 version of SciPy with a relative error close to eps.

jakevdp replied:
Yeah, that's also what the `op_record` test infrastructure does. It's testing with `tol=1E-6`, which is about 8 ULP or so; maybe that's sufficient. 

But you also said that cephes float32 looks very different from float64. That suggests there's something deeper here: do you understand what that is?

jeertmans replied:
I see. My previous tests covered the `[-10, 10]` range, as mentioned in the original docstring (see [here](https://github.com/scipy/scipy/blob/76bf366486b0583dca789823878b80d610eb571b/scipy/special/xsf/cephes/ellik.h#L41-L47)). I have looked at other similar libraries, like TensorFlow, and they only have [one `fresnel_cos` and `fresnel_sin` implementation](https://github.com/tensorflow/tensorflow/blob/52b41bd227b6f280942f04d92076116e62b168e5/tensorflow/core/kernels/special_math/special_math_op_misc_impl.h#L410-L596): the same as the one propose here. If I have understood the code well enough, `Scalar` can either be  `float32` or `float64`, and *I doubt that TensorFlow would automatically promote `float32` to `float64`, compute the Fresnel integrals, and cast the dtype back to `float32`*, as this would be a major memory increase (usually a bottleneck in ML).

Last, I found [that paper](https://link.springer.com/content/pdf/10.1007/s11075-023-01654-2.pdf) (*Calculation of Fresnel integrals of real and complex
arguments up to 28 signifcant digits*, by **Mofreh R. Zaghloul and Leen Alrawas**) that shows the `f` and `g` (eq. 21 and 22) functions that are expanded into series (but the series used are different here, because they aim at a much higher precision).

The difference in power series chosen by Cephes for `float32` and `float64`, that I cannot explain. Except for the smaller amount of computations performed with the `float32` version, that could be justified for the need of less accuracy than the `float64` version.

jakevdp replied:
See my other coment: I think adding test coverage for large values will probably be sufficient.

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,223 @@
+  ])
+
+  @jit
+  def sincospi(
```
you should define jit-compiled helper functions at the outer scope, otherwise this will be re-compiled each time `fresnel` is called, potentially leading to poor performance

jeertmans replied:
Done :-)

jakevdp commented on tests/lax_scipy_special_functions_test.py:
```diff
@@ -95,6 +95,9 @@ def op_record(name, nargs, dtypes, rng_factory, test_grad, nondiff_argnums=(), t
         "factorial", 1, float_dtypes, jtu.rand_default, True
     ),
+    op_record(
+        "fresnel", 1, float_dtypes, jtu.rand_default, True
```
`rand_default` generates normally-distributed values with a standard deviation of 3.

Your function has specific logic for `x > 36974.0` that is untested by this test configuration. I'd suggest adding test coverage of large values, either by changing the rng generator here, or by adding a special test of large values below.

jeertmans replied:
So, the tests were failing for large x values. I implemented the `float32` variant and it now passes the tests. The "bad news" are that the check that `grad(f) = custom_jvp` now fails for `float32` and large input values, as the function returns a constant value of `0.5`, while the derivative is known to be non-zero. I have thus disabled the test on gradient, but I am open to any better solution.

jakevdp commented on jax/_src/test_util.py:
```diff
@@ -773,6 +773,33 @@ def rand_uniform(rng, low=0.0, high=1.0):
   return partial(_rand_dtype, rng.rand, post=post)
 
 
+def rand_fresnel(rng):
```
I don't think we should add a new test utility for a single test.

Why not use `rand_default` with `scale=10000` or something like that?

jeertmans replied:
Good point. This was an easy way for me to make sure to sample points in the three regions, but I can use `rand_default` for sure.

jeertmans commented on tests/lax_scipy_special_functions_test.py:
```diff
@@ -95,8 +95,11 @@ def op_record(name, nargs, dtypes, rng_factory, test_grad, nondiff_argnums=(), t
+    # but the derivative is non-zero.
     op_record(
-        "fresnel", 1, float_dtypes, jtu.rand_fresnel, True
+        "fresnel", 1, float_dtypes, jtu.rand_fresnel, False
```
Yes, I know :/ (See my comment above)

jakevdp replied:
Incorrect gradients seems like a blocker for merging this – any ideas what might be going on?

jeertmans replied:
The issue is that, for single precision, large values of x results in a constant output: 0.5. Thus, the derivative computed with automatic diff is zero. However, the derivative is not zero in reality, and the custom JVP returns the correct derivative (that is trivial).

jakevdp replied:
Are you saying it’s a bug in JAX autodiff rather than a bug in your jvp rule?

jeertmans replied:
Not a bug, but a limitation.

The derivatives of Fresnel integrals are trivial: `cos(x * x * pi / 2)` and `sin(x * x * pi / 2)`. So, instead of relying on an estimate of the derivates through autodiff, I defined a `custom_jvp` function: https://github.com/google/jax/blob/9449e951b58e0d29673165d6cd10b3330454e729/jax/_src/third_party/scipy/special.py#L278-L287

This worked fine with the `float64` implementation of `fresnel`, as the autodiff value would be very close to that of `_fresnel_jvp`.

Now that I have added a special implementation for `float32`, the autodiff value is not close to the value returned by `_fresnel_jvp`. More precisely, when `x` is large, the following lines will return 0 with autodiff:
https://github.com/google/jax/blob/9449e951b58e0d29673165d6cd10b3330454e729/jax/_src/third_party/scipy/special.py#L239-L244

Which differs from the true derivative value (i.e., `_fresnel_jvp`).

jeertmans replied:
Note: it returns 0 because the output value is constant with respect to `x`.

jakevdp replied:
It looks like the problem stems from computing `x * x * pi / 2`. Your implementation avoids this by using `r = jnp.fmod(x, 2.0)` to get the correct phase without being subject to errors from float precision. You could probably use the same strategy in the case of the jvp rule.

jeertmans replied:
I will be off for one week, so I will look at that the week after :)

jeertmans replied:
Hey @jakevdp, so I will try to resume the problem we discussed above.
To recall, the problem is that `jtu.check_grads` fails because the estimated derivative with finite differences does not match the one provided with `custom_jvp`.

I don't think the problem comes from computing `x * x * pi / 2`, as you mentioned, but it *may* increase accuracy to use `sincospi(x)`.

Before b409c6394ea648ef7e9a9a341f859a998440ffd0 and 9449e951b58e0d29673165d6cd10b3330454e729, the tests for `check_grads` worked just fine. The aforementioned commits introduced a new branching for large, single-precision `x` values, where the output is approximately constant, i.e., `c_large = s_large = 0.5`. Hence, estimating the derivative of `fresnel(x)` when `x` is large will always return 0. For double-precision, `fresnel(x)` is not constant but oscillates very rapidly around `0.5`, with a small amplitude (see the above code snippet), so finite-difference may estimate the derivative correctly.

However, we know the derivative of `fresnel(x)` is easy to derive analytically, and this explains why I set up a custom VJP rule. This way, both single-precision and double-precision implementation of `fresnel(x)` have a very accurate estimate of their derivative(s).

As a result, I disabled the test for `check_grads`, but eventually it only needs to be disabled for `abs(x) > 36974.0 and (x.dtype == jnp.float32)`.

What do you think?

jakevdp replied:
> I don't think the problem comes from computing `x * x * pi / 2`

I do think this is the problem: for large `x`, `x * x` is even larger, and when it gets larger than about `10^7`, you can no longer resolve the difference between, say `1` and `1.5`, meaning that you entirely lose the phase information, which is the only relevant thing in the computation.

You can fix this by using modular arithmetic, bounded to the relevant domain of $0\le x < 2$

jeertmans replied:
Ok, I was wrong about my hypothesis, because setting `single_precision = False` should have solved the issue, in theory, but it did not. But now I understand better why you suggested the issue might be in capturing the phase :-)

I quickly tried your suggestion (see 5397fc29578726ca6a449aa2059dc40b650a0f8c), but that doesn't work for the moment (I was quick, so I probably made a mistake), to have `r = x**2 / 2 mod 2 = (x mod 2)(x/2 mod 2) mod 2`.

jeertmans replied:
Ok but the identity I used does not hold, I am dumb haha ^^'

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,287 @@
+
+def _fresnel_jvp(primals, tangents):
+  x, = primals
+  x2 = x * x
```
This is the problematic line: for large `x`, `x * x` in float precision no longer has accuracy to resolve the phase information, so your gradient will always be constant.

Instead we should do something like `r = x % 2` and compute $x^2 \rm{mod} 2$ in terms of this

jakevdp replied:
It looks like this works:
```python
def f(x):
  r = x % 2
  return (r * (x - r / 2)) % 2

np.testing.assert_allclose((x * x / 2) % 2, f(x))
```

jakevdp replied:
We should add a comment explaining this formula; something like this:
```python
# define r = x % 2, y = x - r, then
# (x * x / 2) % 2
#   = [(y + r)*(y + r)/2] % 2
#   = [y*y/2 + r*y + r*r/2] % 2
#   = [(y*y/2%2) + (r*y + r*r/2)%2]%2
#   = [0 + (r*(y+r/2))%2]%2
#   = [r*(x-r/2)]%2
```

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -275,13 +275,41 @@ def fresnel(x: ArrayLike) -> tuple[Array, Array]:
+  Accurate evaluation of sin(pi * x**2 / 2) and cos(pi * x**2 / 2).
+  """
+  x = jnp.abs(x)
+  # x*x/2 mod 2 = ((x mod 2) * (x/2 mod 2)) mod 2
```
This formula is true for integers, but not in general true for real numbers.

jeertmans replied:
Yes good point, thanks!

pearu commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,315 @@
+
+  # Large x values
+
+  sinpi, cospi = sincospi(x2 / 2)
```
```suggestion
  sinpi, cospi = sincospisquaredhalf(x)
```
that ought to resolve https://github.com/google/jax/pull/22843#discussion_r1722119910, wdyt @jakevdp ?

jeertmans replied:
Good point, I do not need `sincospi` anymore then ^^'
But the problem persists.

jakevdp replied:
Which problem persists?

jeertmans replied:
> Which problem persists?

The tests failure on gradients for single precision.

jakevdp replied:
What's the issue now? It seems like we solved the initial problem, which was due to `x*x` losing the phase information.

jeertmans replied:
I hope you don't mind the screenshot 
![image](https://github.com/user-attachments/assets/24363bce-16f2-4dc4-b6f1-36db0404e8da)

The captured stdout comes from an added `print(f"{x = }")` line at the start of the `fresnel` function body.

jakevdp replied:
OK – do you understand why this is going wrong? If we understand it and cannot fix it for fundamental reasons, we should disable the test.

If we don't understand it, we should not disable the test, because it may indicate a bug in the implementation.

Please let me know if you'd like to investigate this, or if you'd like me to fork your branch and take a look.

jeertmans replied:
So I make a small experimental setup that tries to emulate (visually) what `check_grads` does, see the code below.

From the below results (first for `float32`, then for `float64`), I think of the following:

- Using FD as "ground truth" makes little sense when x is large, as the rapid oscillations, with very small amplitude, are hardly possible to capture with FD (changing the `EPS` value does impact the result, but not much);
- For such high values, we will eventually have an issue with single-precision's accuracy, and I don't think we can do anything but acknowledge it (the accuracy of the Fresnel function is usually guaranteed for small values, e.g., `|x| < 5...10`, not for large ones, see the Cephes documentation).

I was a bit surprised that my FD estimate would fail on double-precision too, but variations are very small, so this may explain the "issue" of FD being always zero in that range.

Feel free to explore the code and change the `MIN` value to test different ranges.

Let me know what you think of that :-)

![float32](https://github.com/user-attachments/assets/fb1adea6-cbe9-4b28-a99e-57c680fedd08)

![float64](https://github.com/user-attachments/assets/be706d6c-2bbe-4aa6-87b1-80f679b185b9)



```python
import jax
import jax.numpy as jnp
import jax.scipy.special as jsp
import matplotlib.pyplot as plt
import numpy as np
import scipy.special as sp

dtype = np.float64

MIN = 30000

with jax.experimental.enable_x64(dtype == np.float64):
  x = np.linspace(MIN, MIN + 1, 201, dtype=dtype)
  jx = jnp.asarray(x)
  s, c = sp.fresnel(x)
  js, jc = jsp.fresnel(x)
  djs = jax.vmap(jax.grad(lambda x: jsp.fresnel(x)[0]))(jx)
  djc = jax.vmap(jax.grad(lambda x: jsp.fresnel(x)[1]))(jx)

  fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(15, 10))

  color = "tab:red"
  ax1.set_xlabel('x')
  ax1.set_ylabel("S(x)", color=color)
  ax1.plot(x, s, color=color, label="SciPy")
  ax1.plot(x, js, color="k", linestyle="dotted", alpha=0.5, label="JAX")
  ax1.tick_params(axis="y", labelcolor=color)
  ax1.legend()

  ax1b = ax1.twinx()

  colorb = "tab:blue"
  ax1b.set_ylabel("C(x)", color=colorb)
  ax1b.plot(x, c, color=colorb, label="SciPy")
  ax1b.plot(x, jc, color="k", linestyle="dashed", alpha=0.5, label="JAX")
  ax1b.tick_params(axis="y", labelcolor=colorb)
  ax1b.legend()

  EPS = 1e-3
  s_pos, c_pos = jsp.fresnel(jx + EPS)
  s_neg, c_neg = jsp.fresnel(jx - EPS)

  ds = (s_pos - s_neg) * (0.5 / EPS)
  dc = (c_pos - c_neg) * (0.5 / EPS)

  ax2.set_xlabel('x')
  ax2.set_ylabel("dS(x)", color=color)
  ax2.plot(x, ds, color=color, label="FD")
  ax2.plot(x, djs, color="k", linestyle="dotted", alpha=0.5, label="JAX")
  ax2.tick_params(axis="y", labelcolor=color)
  ax2.legend()

  ax2b = ax2.twinx()

  ax2b.set_ylabel("dC(x)", color=colorb)
  ax2b.plot(x, dc, color=colorb, label="FD")
  ax2b.plot(x, djc, color="k", linestyle="dashed", alpha=0.5, label="JAX")
  ax2b.tick_params(axis="y", labelcolor=colorb)
  ax2b.legend()

  fig.tight_layout()
  plt.show()
```

pearu replied:
Notice that
```python
>>> numpy.nextafter(numpy.float32(30000), numpy.float32(numpy.inf))
30000.002
```
that is, `EPS=1e-3` is too small for single precision large values such as 30000 as `jsp.fresnel(x + EPS)` is very likely equal to `jsp.fresnel(x)` because `x + EPS` is likely equal to `x`. Try `EPS=0.01` but as you said, `jsp.frensel(x)` at the given range is rapidly changing (recall `S(x)->const - cos(x*x)/(2*x) + ...`, `C(x)->const + sin(x*x)/(2*x) + ...` as `x->oo` where `const -> 0.6267..`) so that computing FD may have a little sense indeed. On the other hand, try also
```python
x = np.linspace(MIN, MIN + 0.01, 201, dtype=dtype)
```
to see if you are able to get a smooth graph at all.

Why `dC(MIN)` is equal to 1? Looking at the C(x) graph, I would expect it to be 0.

Also, what is the upper limit for `MIN` to get reasonable results?

jeertmans replied:
> Notice that
> 
> ```python
> >>> numpy.nextafter(numpy.float32(30000), numpy.float32(numpy.inf))
> 30000.002
> ```
> 
> that is, `EPS=1e-3` is too small for single precision large values such as 30000 as `jsp.fresnel(x + EPS)` is very likely equal to `jsp.fresnel(x)` because `x + EPS` is likely equal to `x`. Try `EPS=0.01` but as you said, `jsp.frensel(x)` at the given range is rapidly changing (recall `S(x)->const - cos(x*x)/(2*x) + ...`, `C(x)->const + sin(x*x)/(2*x) + ...` as `x->oo` where `const -> 0.6267..`) so that computing FD may have a little sense indeed.

I agree that `1e-3` may not be appropriate for that range of values, but I chose this value to be the same as the one used by `check_grads`. But there is no best `EPS` value, and using FD might just be bad for this function (for large input values).

> On the other hand, try also
> 
> ```python
> x = np.linspace(MIN, MIN + 0.01, 201, dtype=dtype)
> ```
> 
> to see if you are able to get a smooth graph at all.

![float32_1e-1eps](https://github.com/user-attachments/assets/cd88f5be-5c6a-4d65-b0aa-7eff7247d5b9)

> 
> Why `dC(MIN)` is equal to 1? Looking at the C(x) graph, I would expect it to be 0.
> 
> Also, what is the upper limit for `MIN` to get reasonable results?

Humm good remark, I need to double check that :-)

jakevdp replied:
> * (the accuracy of the Fresnel function is usually guaranteed for small values, e.g., `|x| < 5...10`, not for large ones, see the Cephes documentation).

In that case, we should note this in the function's docstring and ensure that our tests cover the supported range.

jeertmans replied:
> Thanks for digging in! This all makes sense.
> 
> > * (the accuracy of the Fresnel function is usually guaranteed for small values, e.g., `|x| < 5...10`, not for large ones, see the Cephes documentation).
> 
> In that case, we should note this in the function's docstring and ensure that our tests cover the supported range.

Would reducing the scale factor of the random normal variable used for generating test samples be sufficient for you?

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,315 @@
+  """
+  x = jnp.abs(x)
+  r = jnp.fmod(x, 2.0)
+  r = jnp.fmod(r * (x - r / 2), 2.0)
```
We should add a comment explaining this formula; something like this:
```python
# define r = x % 2, y = x - r, then
# (x * x / 2) % 2
#   = [(y + r)*(y + r)/2] % 2
#   = [y*y/2 + r*y + r*r/2] % 2
#   = [(y*y/2)%2 + (r*y + r*r/2)%2]%2
#   = [0 + (r*(y+r/2))%2]%2
#   = [r*(x-r/2)]%2
```

jeertmans commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,293 @@
+  #   = [0 + (r*(y+r/2))%2]%2
+  #   = [r*(x-r/2)]%2
+  r = jnp.fmod(x, 2.0)
+  r = jnp.fmod(r * (x - r / 2), 2.0)
```
Done with 49e537979f4a3ad6e44d49ec7b513e7fd9d104fd :-)

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,293 @@
+  """
+  x = jnp.abs(x)
+  # define s = x % 2, y = x - s, then
+  # (x * x / 2) % 2
```
```suggestion
  # r = (x * x / 2) % 2
```

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,293 @@
+  single_precision = (xxa.dtype == jnp.float32)
+
+  if single_precision:
+    fresnl_sn = jnp.array([
```
The arrays in this block should be explicitly given a `float32` dtype, otherwise they may lead to float64 operations for float32 inputs.

jeertmans replied:
Fixed with 202c726309d9697fd8c25caf408d8e98755c95ba

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,293 @@
+
+  isinf = jnp.isinf(xxa)
+  small = x2 < 2.5625
+  large = x > 36974.0
```
This is what is confusing to me: we have explicit logic for `x > 36974`, but you mentioned the docs say the outputs are only good for $|x| < 10$. How do I reconcile those two things?

jeertmans replied:
Well I did not claim (nor they did) that it would fail for large values, but reading the source code, this is what they say

>  * ACCURACY:
>  *
>  *  Relative error.
>  *
>  * Arithmetic  function   domain     # trials      peak         rms
>  *   IEEE       S(x)      0, 10       30000       1.1e-6      1.9e-7
>  *   IEEE       C(x)      0, 10       30000       1.1e-6      2.0e-7
>  */

for the single-precision version, and

>  * ACCURACY:
>  *
>  *  Relative error.
>  *
>  * Arithmetic  function   domain     # trials      peak         rms
>  *   IEEE       S(x)      0, 10       10000       2.0e-15     3.2e-16
>  *   IEEE       C(x)      0, 10       10000       1.8e-15     3.3e-16
>  *   DEC        S(x)      0, 10        6000       2.2e-16     3.9e-17
>  *   DEC        C(x)      0, 10        5000       2.3e-16     3.9e-17
>  */

for the double-precision version.

So while they have specific cases for large values, I don't know what one can expect regarding their accuracy. I could not find more documentation about them than that of https://www.netlib.org/cephes/.

jakevdp replied:
OK, I guess I misunderstood your earlier comment in that case.

All the details aside, what I'm saying is this: we should document and test the range of values over which we expect this function to be accurate.

jeertmans replied:
I added some comment about this, but I cannot find a range that works (where previously I swear it would), but I cannot spend more time on this PR at the moment :/

jakevdp replied:
There's still some work to be done to land this – are you saying you no longer want to contribute this?

jeertmans replied:
Not that I don't want to contribute to this, but I will need a pause on this for at least one week, and continue later (probably after 2nd of September). If anyone is willing to give some help, I'd gladly take some :-)

jakevdp replied:
Sure, sounds good!

jeertmans replied:
Hello @jakevdp! So I found out why tests were not working anymore, a dumb typo felt in the Cephes coefficients for single precision, but it's now fixed. I have adjusted the scale so that it now test a range we are sure the coefficients were developed for. It is documented in the docstring. For the half-float dtypes, see the other comments thread :-)

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,293 @@
+  sinpi, cospi = sincospisquaredhalf(x)
+  dSdx = sinpi * x_dot
+  dCdx = cospi * x_dot
+  return result, (dSdx.astype(result[0].dtype), dCdx.astype(result[0].dtype))
```
I suspect these typecasts will no longer be necessary once the `dtype` of the arrays above is fixed.

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,299 @@
+
+  # This part is mostly a direct translation of SciPy's C++ code.
+
+  single_precision = (xxa.dtype == jnp.float32)
```
We need to decide what to do with float16 and bfloat16 inputs. As currently written, the implementation will effectively promote them to whatever the widest available float type is (float64 if enable_x64 is True, float32 if not). Generally we would want the output dtype to match the input dtype

jakevdp replied:
Take a look at how `bessel_i0e_impl` handles various input dtypes – I think we should use the same pattern here.

jeertmans replied:
> We need to decide what to do with float16 and bfloat16 inputs. As currently written, the implementation will as a side-effect promote them to whatever the widest available float type is (float64 if enable_x64 is True, float32 if not). Generally we would want the output dtype to match the input dtype

I totally forgot they existed, good point. Well, upcasting is the easy solution (as what they do with `bessel_i0e_impl`). But, before that, I think it would be important to check if using, e.g., the `float32` coefficients for `float16` results in accurate results, or not.

jeertmans replied:
After some tests, it appears that `float16` overwflows for any value in the "other" range, which means the values are NaNs for x > 3 (roughly). For `bfloat16`, series expansions start to diverge for x > 8~9, which is better but still not very good.

Based on the Bessel example, I decided to manually upcast the arrays elements to `float32`, and downcast them if needed.

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,320 @@
+      +9.999841934744914e-1,
+    ], dtype=dtype)
+
+  else:
```
This should be `elif dtype == float64`

jeertmans replied:
I did not know JAX supported float8 😅

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,320 @@
+      +8.39158816283118707363e-19,
+      +1.86958710162783236342e-22,
+    ], dtype=dtype)
+
```
You should add an `else:` statement here and raise an appropriate error, otherwise input types that you'd not handled (e.g. `float8`) will fail in confusing ways (probably a `NameError` on `fresnel_sn`).

jeertmans replied:
But they shouldn’t I upcast too?

jakevdp replied:
We could upcast float8, but we still should have an `else` block that raises an error for unexpected inputs.

jakevdp replied:
Actually, it’s probably better to error for float8 inputs.

jeertmans replied:
Let me know if b0e4a6a22d7c178d771bc0fa2e91ad6beffd466c addresses your comment.

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,324 @@
+
+  single_precision = (dtype == jnp.float32)
+
+  if single_precision:
```
This may seem nitpicky, but handling the logic of dtype support with two different if-else blocks will make this code harder to maintain in the long run. A future reader of this code trying to debug an issue with, say, a new dtype, will have to parse by eye several dozen lines of code & understand the provenance of several related variables (`single_precision`, `dtype`, and `original_dtype`) in order to understand how the logic depends on the input dtype. It would be much more immediately understandable to put all this logic in one place. For example:
```python
if dtypes.issubdtype(xxa.dtype, dtypes.complexfloating):
  raise ValueError("complex values not supported")
elif xxa.dtype == jnp.float64:
  # define double-precision parameters
elif xxa.dtype in (jnp.float32, jnp.float16, jnp.bfloat16):
  xxa = xxa.astype('float32')
  # define single-precision paramters
else:
  raise ValueError(f"unsuppored dtype: {dtype}")

assert xxa.dtype in ('float32', 'float64')
single_precision = (xxa.dtype == 'float32')
```
What do you think?

jeertmans replied:
I agree this could be simpler, what do you think of d69866043e9731275405ac1eb33811c3ad843bb5 ?

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,325 @@
+    # For half-precision, series expansions have either
+    # produce overflow or poor accuracy.
+    # Upcasting to single-precision is hence needed.
+    original_dtype = xxa.dtype
```
The `original_dtype` definition could probably go above the `if` block, because it will always be the same.

jakevdp commented on jax/_src/third_party/scipy/special.py:
```diff
@@ -0,0 +1,325 @@
+      +7.244727626597022e-3,
+      -9.228055941124598e-2,
+      +5.235987735681432e-1,
+    ], dtype=xxa.dtype)
```
Slight preference here and below to just explicitly say `dtype='jnp.float32`, since this will always be the case and its clearer than having to check what `xxa.dtype` is after all the typecasting.

jeertmans replied:
Done!

jakevdp reviewed (commented):
Two last small comments, but this looks great!

Once you make the last changes, please squash all changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests). This kind of squashing & force-pushing workflow is not all that common, so feel free to reach out if you need help with it.

Thanks for sticking with this PR through the many rounds of review – it will be a great addition to the JAX package!

jeertmans committed:
feat(lib): add real-valued implementation of jax.scipy.special.fresnel

Add implementation, documentation, and tests, for both single-precision and double-precision floating-point arithmetic.

jeertmans commented:
> Two last small comments, but this looks great!
> 
> Once you make the last changes, please squash all changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests). This kind of squashing & force-pushing workflow is not all that common, so feel free to reach out if you need help with it.
> 
> Thanks for sticking with this PR through the many rounds of review – it will be a great addition to the JAX package!

Thanks for your review!

I did squash all the commits into one, so it should be all good on my side. I also fixed the unused import that was left. Let me know if there is anything more I should do!

This function is rather essential to my research, so I will be happy to remove all that code from my code-base, and only have to use `jax.scipy.special.fresnel` instead :D

jeertmans commented:
Hi @hawkinsp, I saw you just released [`jax-v0.4.32`](https://github.com/google/jax/releases/tag/jax-v0.4.32), but I could not see any mention to the new `jax.scipy.fresnel` function in the changelog, is it on purpose?

I am of course biased as it this is my PR, but I was hoping it could be listed in the changelog, so other people might be aware that the function is now available :-)
